### DISCUSSION TITLE
Dynamic content based on production/development enviorment

### DISCUSSION THREAD

**Discussion by joapuiib** [mkdocs/mkdocs#3768] [Q&A] [answered]
Hello,
I would like to know if Mkdocs provides a way to check in which environment is active (`development` or `production`). In other words, I would link to know if mkdocs is `serving` or `building`.

I want to enable a comment system (provided by [`mkdocs-material`](https://squidfunk.github.io/mkdocs-material/setup/adding-a-comment-system/?h=comme)) only when the site is built, but not in development.

I'm looking for something similar to https://middlemanapp.com/basics/helper-methods/#environment-helpers

```html
<!-- TODO: Render content if building site -->
{% if page.meta.comments and mkdocs.build %}
  <h2 id="__comments">{{ lang.t("meta.comments") }}</h2>
  <!-- Insert generated snippet here -->

  <!-- Synchronize Giscus theme with palette -->
  <script>
    var giscus = document.querySelector("script[src*=giscus]")

    // Set palette on initial load
    var palette = __md_get("__palette")
    if (palette && typeof palette.color === "object") {
      var theme = palette.color.scheme === "slate"
        ? "transparent_dark"
        : "light"

      // Instruct Giscus to set theme
      giscus.setAttribute("data-theme", theme) 
    }

    // Register event handlers after documented loaded
    document.addEventListener("DOMContentLoaded", function() {
      var ref = document.querySelector("[data-md-component=palette]")
      ref.addEventListener("change", function() {
        var palette = __md_get("__palette")
        if (palette && typeof palette.color === "object") {
          var theme = palette.color.scheme === "slate"
            ? "transparent_dark"
            : "light"

          // Instruct Giscus to change theme
          var frame = document.querySelector(".giscus-frame")
          frame.contentWindow.postMessage(
            { giscus: { setConfig: { theme } } },
            "https://giscus.app"
          )
        }
      })
    })
  </script>
{% endif %}
```

**Comment by pawamoy** [comment]
You could set an `extra` value from an environment variable using `!ENV` in mkdocs.yml. You can then access this extra variable in the template. I'm on mobile so will let you check the docs :)

**Reply by joapuiib** [reply]
I'm looking for something similar to https://middlemanapp.com/basics/helper-methods/#environment-helpers, so I don't have to modify a value (or set at environment variable).

Is it possible or should I use a `!ENV` var?

**Comment by kamilkrzyskow** [comment, marked as answer]
The global context doesn't provide anything related to the serve/build mode @joapuiib:
- https://www.mkdocs.org/dev-guide/themes/#global-context

You could make a simple hook to fix that:
- https://www.mkdocs.org/user-guide/configuration/#hooks
- https://www.mkdocs.org/dev-guide/plugins/#events

Use a global variable and set it in the `on_startup` event then in `on_page_context` modify the context to add the value set in `on_startup`.
Let me know if you require more help :v:

**Reply by pawamoy** [reply]
I was going to mention hooks next, you beat me to it! Good answer :+1:

**Reply by joapuiib** [reply]
I've written a plugin with that idea and it works! Thanks!

I'll leave it here, just in case it helps somebody:

```py
from mkdocs.plugins import BasePlugin

class EnviormentPlugin(BasePlugin):

    def __init__(self):
        self.enabled = True
        self.total_time = 0
        self.is_dirty = False

        self.is_building = True

    def on_startup(self, *, command, dirty):
        self.is_dirty = dirty

    def on_serve(self, server, config, builder):
        self.is_building = False
        return server

    def on_page_context(self, context, page, config, nav):
        context['build'] = self.is_building
        return context
```

Then, I can use it in any page.
```
{% if page.meta.comments and page.build %}
    ...
{% endif %}
```

**Reply by kamilkrzyskow** [reply]
Quite a bit of redundancy here, like unused variables or the `on_startup` event should be enough to decide if `is_building` is True/False.
Also I think that `page.build` should simply be a `build`, as you set the `build` in the context.
I still feel like a hook would be simpler, as you don't need to setup a plugin with an entrypoint.

Nevertheless, thanks for sharing 🤩

**Reply by joapuiib** [reply]
You are right. It was not working properly with `page.build`.

I've updated my plugin with your new recommendations. Thanks for reviewing the code! :smile: 

```py
from mkdocs.plugins import BasePlugin

class EnviormentPlugin(BasePlugin):

    def __init__(self):
        self.enabled = True
        self.is_building = False

    def on_startup(self, *, command, dirty):
        self.is_building = command in ['build', 'gh-deploy']

    def on_page_context(self, context, page, config, nav):
        context['build'] = self.is_building
        return context
```

And then:

```py
{% if page.meta.comments and build %}
    ...
{% endif %}
```

I've chosen to write a plugin, so I can include it in a custom theme and use it in all my mkdocs subprojects.

**Comment by codeofandrin** [comment]
As @pawamoy suggested I just used the `extra` fields in `mkdocs.yml` and then used the `on_env` hook to add it to the global env.

`mkdocs.yml`:
```yml
extra:
    is_prod: !ENV [CI, false]  # in my case, if it's not generated via CI, it's served locally
hooks:
  - path/to/hooks.py
```

`hooks.py`:
```py
import mkdocs

@mkdocs.plugins.event_priority(-100)
def on_env(env, config, files):
    env.globals["is_dev"] = not config.extra["is_prod"]

    return env
```
